### ISSUE TITLE
Merge Key Language-Independent Type for YAML Support

### ISSUE THREAD

**Issue by DavidChennells** [cloudbase/powershell-yaml#33]
Does this project support the Merge Key Language-Independent Type for YAML Version 1.1?

I'm encountering the same behavior in powershell-yaml as the one reported and fixed in the lyaml library described here:  [YAML merge key not working](https://github.com/gvvaughan/lyaml/issues/8#issue-94299178)

Within the learnxinyminutes.yaml example, access to the bar.name property seems to work only via:

$yaml.bar.'<<'.name

Is this the intended model?

**Comment by gabriel-samfira** [comment]
Looks like this is something that YamlDotNET supports. I will look into adding this to powershell-yaml.

> $yaml.bar.'<<'.name
> Is this the intended model?

No, definitely not. It should be seamless. Keys should be inherited correctly. I may have to update the assemblies, and expose some new functionality to get the desired outcome. I will look into this as soon as I can spare some time. No ETA though.

**Comment by gabriel-samfira** [comment]
Would you mind giving this branch a try:

https://github.com/cloudbase/powershell-yaml/tree/use-merging-parser

And see if this works for you? If it does, I will merge this into master and publish on powershell gallery, as soon as I can write some tests for it.

**Comment by DavidChennells** [comment]
Many thanks for the immediate response!

The single map and multiple map cases, which are the most important, are working great in that branch.

However, the override case, which is also envisaged in the [YAML 1.1 Spec for Merge Keys](http://yaml.org/type/merge.html), is throwing a "Duplicate key" error.

The stack trace for the error points to $yamlStream.Load([YamlDotNet.Core.IParser] $merging).

Not sure whether improving the duplicate error key detection logic for merging in order to resolve this problem will require coordinating with the YamlDotNET project.

**Comment by gabriel-samfira** [comment]
This will indeed have to be raised in the YamlDotNet package, and I think this can be confined to the merging parser. But I am not deeply familiar with the YamlDotNet codebase. I tried the latest version of the assemblies, and this issue seems to be present there as well.

I can write tests for the cases that do work, and update the package on powershellgallery sometime this week (or in the weekend). This way, you will at least be partially unblocked from using the module until the overwrite case is treated by YamlDotNet. Does that sound acceptable for the moment?

**Comment by DavidChennells** [comment]
Sure that's fabulous.  Will you include a test for the override case (which it appears will fail), perhaps in a branch, and raise the issue with the other team once you've independently confirmed my observation?

**Comment by DavidChennells** [comment]
Thanks very much Gabriel.  Ideally, in the future YamlDotNet would handle the selection of the parser internally (or unify their parsers into one per non-backwards-compatible Yaml specification).  It is not ideal for clients to handle details of parser selection, especially given that breaking changes to existing scripts are not very likely with the use of the special ">>" string.

Another small note:  I believe "node" should read "note" in the text you've added to the ReadMe.md.

**Comment by gabriel-samfira** [comment]
Indeed, ideally, the merging parser should be used by default. Until the override case is treated, I don't feel comfortable using it by default. Even so, I am weary of changing behavior from one minor version to another, even if it is to correct wrong behavior.

When/if the merging parser (actually the YamlStream class) will fully treat the override case, I will glagly switch to that as the default.

I agree with you, there should be no need to pick one or the other to begin with, and simply have one parser.
